ui/client/components/aiAssistant/AIAssistant.js

```diff
@@ -71,67 +79,96 @@ const AIAssistant = () => {
+        } catch (e) {
+          console.log('error fetching document, skipping', e);
+          return { filename: null, data: {} };
+        }
+      });
+
+      // wait for all the document fetches to complete
+      const documentResults = await Promise.all(documentFetches);
+      // and map them to the filenames
+      documentResults.forEach(({ filename, data }) => {
+        filenamesToDocs[filename] = data;
+      });
+
+      response.documents = filenamesToDocs;
+
+      setResponses((oldResponses) => ({
+        ...oldResponses,
+        [queryKey]: {
+          data: response,
+          status: 'success',
+        }
+      }));
+    } catch (error) {
+      console.error(error);
+      setResponses((oldResponses) => ({
+        ...oldResponses,
+        [queryKey]: {
+          data: null,
+          status: 'error',
+        }
+      }));
+    } finally {
+      setAnyResponseLoading(false);
+    }
+  };
 
   const handleSearch = async () => {
     if (searchPhrase.length) {
-      // store all details for this query in this object: question, response, and documents
-      const queryDetails = { question: searchPhrase };
-      try {
-        // TODO: Use this for local development, causemos-analyst for production
-        // const queryResp = await axios.get(`http://localhost:8001/mock-message?query=${searchPhrase}`);
-
-        const mock = ''; //'mock-';
-
-        const queryResp = await axios.get(
-          `/api/ai-docs/${mock}message?query=${searchPhrase}`
-        );
-        queryDetails.response = queryResp.data;
-
-        // create an object with just the unique filenames as keys
-        const filenamesToDocs = queryResp.data.candidate_paragraphs.reduce((obj, curr) => (
-          { ...obj, [curr.root_name]: null }
-        ), {});
-
-        // go through the unique filenames and fetch the document data
-        const documentFetches = Object.keys(filenamesToDocs).map(async (filename) => {
-          try {
-          const docFetchResp = await axios.get(`/api/dojo/documents/by-didx-name?name=${filename}`);
-            return { filename, data: docFetchResp.data };
-          } catch (e) {
-            console.log('error fetching document, skipping', e);
-            return {filename: null, data: {}};
-          }
-        });
-
-        // wait for all the document fetches to complete
-        const documentResults = await Promise.all(documentFetches);
-        // and map them to the filenames
-        documentResults.forEach(({ filename, data }) => {
-          filenamesToDocs[filename] = data;
-        });
-
-        queryDetails.documents = filenamesToDocs;
-
-        setPreviousSearches((oldPairs) => ([...oldPairs, queryDetails]));
-        setSearchPhrase('');
-      } catch (error) {
-        console.error(error);
-        // Handle errors
-      }
+      // store the user input query and the responses keyed under a timestamp
+      // so we can match them up
+      const queryKey = Date.now();
```

Comment:
Queries and responses are now broken apart (and matched up by the queryKey) so that we can save the query and display it right away even if we don't get a response. `performSearch` is broken out into its own function above so that we can try fetching the response again (and fill out just the response object under a given `queryKey`) if we get an error and the user clicks the "Try Again" button to retry their query.